src/libos/src/net/socket/unix/stream/endpoint.rs

```diff
@@ -100,6 +103,42 @@ impl Inner {
         Ok(())
     }
 
+    pub fn poll(&self) -> IoEvents {
+        let mut events = IoEvents::empty();
+
+        if self.reader.is_self_shutdown() || self.writer.is_self_shutdown() {
+            return IoEvents::RDHUP | IoEvents::HUP;
+        }
+
+        // get the status of peer
+        if self.reader.is_peer_shutdown() {
+            events |= IoEvents::RDHUP;
+            if self.writer.is_peer_shutdown() {
+                events |= IoEvents::HUP;
+            }
+        }
+
+        let reader_events = self.reader.poll();
+        let writer_events = self.writer.poll();
+        events |= (reader_events & IoEvents::IN) | (writer_events & IoEvents::OUT);
```

Comment:
Needs experimental results to confirm that the event conversion is consistent with Linux's behavior.